src/api/defense/util.rs

```diff
@@ -379,47 +380,55 @@ pub fn set_map_invalid(conn: &mut PgConnection, map_id: i32) -> Result<()> {
     Ok(())
 }
 
-pub fn fetch_defense_history(
-    defender_id: i32,
+pub fn fetch_defense_historyboard(
     user_id: i32,
+    page: i64,
+    limit: i64,
     conn: &mut PgConnection,
-) -> Result<GameHistoryResponse> {
+) -> Result<HistoryboardResponse> {
     use crate::schema::{game, levels_fixture, map_layout};
 
     let joined_table = game::table.inner_join(map_layout::table.inner_join(levels_fixture::table));
-    let games_result: Result<Vec<GameHistoryEntry>> = joined_table
-        .filter(game::defend_id.eq(defender_id))
+
+    let total_entries: i64 = joined_table
+        .filter(game::defend_id.eq(user_id))
+        .count()
+        .get_result(conn)
+        .map_err(|err| DieselError {
+            table: "game",
+            function: function!(),
+            error: err,
+        })?;
+    let off_set: i64 = (page - 1) * limit;
+    let last_page: i64 = (total_entries as f64 / limit as f64).ceil() as i64;
+
+    let games_result: Result<Vec<HistoryboardEntry>> = joined_table
+        .filter(game::defend_id.eq(user_id))
+        .offset(off_set)
+        .limit(limit)
         .load::<(Game, (MapLayout, LevelsFixture))>(conn)
         .map_err(|err| DieselError {
             table: "game",
             function: function!(),
             error: err,
         })?
         .into_iter()
-        .map(|(game, (_, levels_fixture))| {
+        .map(|(game, (_map_layout, levels_fixture))| {
             let is_replay_available = api::util::can_show_replay(user_id, &game, &levels_fixture);
-            let attacker = fetch_user(conn, game.attack_id)?.ok_or(AuthError::UserNotFound)?;
-            let defender = fetch_user(conn, game.defend_id)?.ok_or(AuthError::UserNotFound)?;
-            Ok(GameHistoryEntry {
-                game,
-                attacker: UserDetail {
-                    user_id: attacker.id,
-                    username: attacker.username,
-                    trophies: attacker.trophies,
-                    avatar_id: attacker.avatar_id,
-                },
-                defender: UserDetail {
-                    user_id: defender.id,
-                    username: defender.username,
-                    trophies: defender.trophies,
-                    avatar_id: defender.avatar_id,
-                },
-                is_replay_available,
+            Ok(HistoryboardEntry {
+                opponent_user_id: game.attack_id,
+                is_attack: false,
+                damage_percent: game.damage_done,
+                result: "none".to_string(),
+                artifacts_taken: game.artifacts_collected,
```

Comment:
Shouldn't this be -ve, cuz it is the number of artifacts lost

Reply:
You are right!